lib/internal/dns.js

```diff
@@ -0,0 +1,2624 @@
+    if (answers.length === 0) {
+      // Emulate old DNS behavior of sending an Error on "empty" responses
+      // instead of sending an empty array like with dns.resolve*()
+      return resolvecb(errnoException('ENOTFOUND', hostname));
+    }
+
+    if (!isCustomType) {
+      var needv4Mapped = (v4mapped && family === 6 && triedIPv4);
+      var addr;
+      if (family === 0)
+        family = (triedIPv4 ? 4 : 6);
+      if (all) {
+        var addrs = new Array(answers.length);
+        for (var i = 0; i < answers.length; ++i) {
+          addr = answers[i].data;
+          if (needv4Mapped)
+            addr = '::ffff:' + addr;
+          addrs[i] = {
+            address: addr,
+            family: family
+          };
+        }
+        callback(null, addrs);
+      } else {
+        addr = answers[0].data;
+        if (needv4Mapped)
+          addr = '::ffff:' + addr;
+        callback(null, addr, family);
+      }
+    } else {
+      if (all) {
+        var vals = new Array(answers.length);
+        for (var i = 0; i < answers.length; ++i)
+          vals[i] = answers[i].data;
+        callback(null, vals);
+      } else {
+        callback(null, answers[0].data);
+      }
+    }
+  });
+};
+
+function resolver(type) {
+  var returnFirst = (type === 'SOA');
+  return function query(name, callback) {
+    if (typeof name !== 'string')
+      throw new Error('Name must be a string');
+    else if (typeof callback !== 'function')
+      throw new Error('Callback must be a function');
+
+    resolve(name, type, FLAG_RESOLVE_NSONLY, function resolvecb(err, rep) {
+      if (err)
+        return callback(err);
+
+      var answers = rep.answers;
+      var addrs;
+
+      if (returnFirst)
+        addrs = (answers.length ? answers[0].data : null);
+      else {
+        addrs = new Array(answers.length);
+        for (var i = 0; i < answers.length; ++i)
+          addrs[i] = answers[i].data;
+      }
+
+      callback(null, addrs);
+    });
+  };
+}
+
+var resolveMap = {};
+(function() {
+  var keys = Object.keys(QTYPE_V2O);
+  for (var i = 0; i < keys.length; ++i) {
+    var key = keys[i];
+    var name = QTYPE_V2O[key].name;
+    var exportName = 'resolve' + name[0] + name.slice(1).toLowerCase();
+    if (name === 'A')
+      exportName = 'resolve4';
+    else if (name === 'AAAA')
+      exportName = 'resolve6';
+    else if (name === 'PTR')
+      exportName = 'reverse';
+    exports[exportName] = resolveMap[name] = resolver(name);
+  }
+  exports.resolveAll = resolveMap['*'] = resolver('*');
+})();
+
+exports.resolve = function(hostname, type, callback_) {
+  var resolver;
+  var callback;
+  if (typeof type === 'string') {
+    resolver = resolveMap[type];
+    callback = callback_;
+  } else if (typeof type === 'function') {
+    resolver = exports.resolve4;
+    callback = type;
+  } else {
+    throw new Error('Type must be a string');
+  }
+
+  if (typeof resolver === 'function')
+    resolver(hostname, callback);
```

Comment:
Well it may already be a `semver-major` because it's replacing a lot of machinery. At least I thought that was the case for the replacement of the `url` module and the http parser.